File: sdk/go/config_test.go
package shojiku

import (
	"context"
	"slices"
	"strings"
	"testing"
)

func TestAnExplicitOptionBeatsConfigureWhichBeatsTheEnvironment(t *testing.T) {
	t.Setenv("SHOJIKU_TEMPLATE_ROOT", "/from-env")
	t.Setenv("SHOJIKU_BIN", engineBinary(t))
	t.Cleanup(ResetConfiguration)

	fromEnv, err := NewClient()
	if err != nil {
		t.Fatalf("building: %v", err)
	}
	if fromEnv.TemplateRoot() != "/from-env" {
		t.Errorf("root = %q, want the environment's", fromEnv.TemplateRoot())
	}

	Configure(WithTemplates("/from-configure"))
	configured, err := NewClient()
	if err != nil {
		t.Fatalf("building: %v", err)
	}
	if configured.TemplateRoot() != "/from-configure" {
		t.Errorf("root = %q, want Configure's", configured.TemplateRoot())
	}

	explicit, err := NewClient(WithTemplates("/explicit"))
	if err != nil {
		t.Fatalf("building: %v", err)
	}
	if explicit.TemplateRoot() != "/explicit" {
		t.Errorf("root = %q, want the explicit option's", explicit.TemplateRoot())
	}
}

func TestTheBinaryResolvesTheOtherWayRound(t *testing.T) {
	// The deliberate asymmetry: WHERE THE ENGINE LIVES is a deployment
	// decision that has to be able to win over application code, while WHICH
	// TEMPLATES an application renders is the application's own.
	installed := engineBinary(t)
	t.Setenv("SHOJIKU_BIN", installed)
	t.Cleanup(ResetConfiguration)

	client, err := NewClient(WithBinary("/some/other/shojiku"))

	if err != nil {
		t.Fatalf("building: %v", err)
	}
	if got := client.settings.engine.binary.path; got != installed {
		t.Errorf("binary = %q, want the environment's %q", got, installed)
	}
	if got := client.settings.engine.binary.source; got != "environment" {
		t.Errorf("source = %q, want %q", got, "environment")
	}
}

func TestStrictIsOredRatherThanOverridden(t *testing.T) {
	base := config{}

	on := base.merge([]Option{WithStrict(true)})
	if !on.strict {
		t.Error("WithStrict(true) did not take")
	}
	if lifted := on.merge([]Option{WithStrict(false)}); !lifted.strict {
		t.Error("a declared lockdown was lifted by a later option")
	}
	if never := base.merge([]Option{WithStrict(false)}); never.strict {
		t.Error("WithStrict(false) turned strictness on")
	}
}

func TestProvidersReplaceRatherThanMerge(t *testing.T) {
	// A client that declares its own registry is stating the whole set it may
	// sign with; quietly adding globally-registered keys would defeat that.
	global := testSigner(t)
	own := testSigner(t)
	Configure(WithProviders(map[string]*LocalPem{"global": global}))
	t.Cleanup(ResetConfiguration)

	merged := globalDefaults().merge([]Option{WithProviders(map[string]*LocalPem{"own": own})})

	if _, present := merged.providers["global"]; present {
		t.Error("the globally-registered provider survived into a client's own registry")
	}
	if merged.providers["own"] != own {
		t.Error("the client's own provider is missing")
	}
}

func TestARegisteredProviderMapIsCopiedRatherThanAliased(t *testing.T) {
	registry := map[string]*LocalPem{"release": testSigner(t)}

	merged := config{}.merge([]Option{WithProviders(registry)})
	delete(registry, "release")

	if merged.providers["release"] == nil {
		t.Error("mutating the caller's map changed the client's registry")
	}
}

func TestResetConfigurationDropsEveryDefault(t *testing.T) {
	// Public because a global that cannot be reset makes every test suite
	// invent its own teardown — and get it wrong.
	Configure(WithTemplates("/somewhere"), WithLang("ja-JP"), WithStrict(true))

	ResetConfiguration()

	if got := globalDefaults(); got.templates != nil || got.lang != nil || got.strict {
		t.Errorf("defaults survived the reset: %+v", got)
	}
}

func TestAPerCallLocaleBeatsTheClientWideOne(t *testing.T) {
	// Go's params are an ordinary argument, so the frozen rule applies
	// directly here rather than through a derived client. What the seven
	// mirror is the precedence, not the spelling.
	client := newTestClient(t, WithLang("en-US"))

	if got := firstNonEmpty(resolveCall([]CallOption{Lang("ja-JP")}).lang, client.settings.lang()); got != "ja-JP" {
		t.Errorf("lang = %q, want the per-call value", got)
	}
	if got := firstNonEmpty(resolveCall(nil).lang, client.settings.lang()); got != "en-US" {
		t.Errorf("lang = %q, want the client-wide value", got)
	}
}

func TestAPerCallLocaleReachesTheEngine(t *testing.T) {
	argvLog := t.TempDir() + "/argv"
	client := stubClient(t, stubBinary(t, `for arg in "$@"; do printf '%s\n' "$arg" >> `+
		argvLog+`; done
printf '{"ok":true,"diagnostics":{"items":[]}}' > "$report"
exit 0`), WithLang("en-US"))

	if _, err := client.Generate(context.Background(), "receipt",
		map[string]any{}, Lang("ja-JP")); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	argv := readLines(t, argvLog)
	if !slices.Contains(argv, "ja-JP") {
		t.Errorf("the per-call locale did not cross:\n%s", strings.Join(argv, "\n"))
	}
	if slices.Contains(argv, "en-US") {
		t.Errorf("the client-wide locale crossed instead:\n%s", strings.Join(argv, "\n"))
	}
}

func TestNoLocaleAtAllPassesNoFlag(t *testing.T) {
	argvLog := t.TempDir() + "/argv"
	client := stubClient(t, stubBinary(t, `for arg in "$@"; do printf '%s\n' "$arg" >> `+
		argvLog+`; done
printf '{"ok":true,"diagnostics":{"items":[]}}' > "$report"
exit 0`))

	if _, err := client.Generate(context.Background(), "receipt", map[string]any{}); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if slices.Contains(readLines(t, argvLog), "--lang") {
		t.Error("a locale flag crossed for a client that has none")
	}
}

func TestNoMemoizedDefaultClientShips(t *testing.T) {
	// A package-level singleton would add a reset-on-reconfigure lifecycle
	// that seven languages would each get subtly wrong. Two clients built
	// from the same configuration are two clients.
	Configure(WithTemplates(fixtureTemplates(t)), WithBinary(engineBinary(t)))
	t.Cleanup(ResetConfiguration)

	first, err := NewClient(WithEnv(false))
	if err != nil {
		t.Fatalf("building: %v", err)
	}
	second, err := NewClient(WithEnv(false))
	if err != nil {
		t.Fatalf("building: %v", err)
	}

	if first == second {
		t.Error("two constructions returned one client")
	}
}

func TestBuildingAClientOverAMissingEngineFailsAtConstruction(t *testing.T) {
	// The reference opens its engine in the constructor so a container learns
	// at boot that the engine is not installed, rather than at the first
	// request.
	_, err := NewClient(WithEnv(false), WithBinary("/no/such/shojiku"))

	if err == nil {
		t.Fatal("a client was built over a binary that does not exist")
	}
	var notFound *BinaryNotFoundError
	if !asError(err, &notFound) {
		t.Fatalf("err = %v, want a BinaryNotFoundError", err)
	}
}

File: sdk/java/src/test/java/jp/kengos/shojiku/ConfigTest.java
package jp.kengos.shojiku;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/**
 * Precedence: what an explicit argument beats, what configuration beats, and the two places the
 * order deliberately reverses.
 */
class ConfigTest extends Fixtures {

  @Test
  void configureSetsProcessWideDefaults() {
    Configuration.configure(config -> config.withTemplates("/configured"));

    assertEquals("/configured", Configuration.current().templates());
  }

  @Test
  void anExplicitArgumentBeatsAConfiguredDefault() {
    // Configuration feeds the same builder; it never adds a precedence level of
    // its own.
    Configuration.configure(config -> config.withTemplates("/configured"));

    assertEquals(TEMPLATES, client().build().templateRootOrNull().path());
  }

  @Test
  void anAbsentArgumentInheritsTheConfiguredDefault() {
    Configuration.configure(config -> config.withTemplates(TEMPLATES).withLang("ja-JP"));

    assertEquals(TEMPLATES, rootless().build().templateRootOrNull().path());
  }

  @Test
  void strictIsTheOnePlaceConfigurationBeatsTheCallSite() {
    // A restriction an operator declared must not be liftable by application code:
    // `strict` is OR-ed rather than overridden.
    assertTrue(Config.empty().withStrict(true).merge(Config.empty().withStrict(false)).isStrict());
  }

  @Test
  void aCallSiteCanStillTightenWhatConfigurationLeftOpen() {
    assertTrue(Config.empty().withStrict(false).merge(Config.empty().withStrict(true)).isStrict());
  }

  @Test
  void providersReplaceRatherThanMerge() {
    // A client that declares its own registry is stating the whole set it may sign
    // with; quietly adding globally-registered keys would defeat the point.
    Map<String, SigningProvider> global = Map.of("global", signer());
    Map<String, SigningProvider> local = Map.of("local", signer());

    Config merged = Config.empty().withProviders(global).merge(Config.empty().withProviders(local));

    assertEquals(java.util.Set.of("local"), merged.providers().keySet());
  }

  @Test
  void aClientThatDeclaresNoRegistryInheritsTheConfiguredOne() {
    Configuration.configure(config -> config.withProviders(Map.of("release", signer())));

    ShojikuClient client = client().build();

    assertTrue(client.sign(rendered(client), "release").success());
  }

  @Test
  void resetDropsEveryConfiguredDefault() {
    // Public because a global that cannot be reset makes every test suite invent
    // its own teardown — and get it wrong in a randomly-ordered run.
    Configuration.configure(config -> config.withTemplates("/configured").withStrict(true));

    Configuration.reset();

    assertNull(Configuration.current().templates());
    assertFalse(Configuration.current().isStrict());
    assertTrue(Configuration.current().envEnabled());
  }

  @Test
  void everySettingSurvivesTheMerge() {
    ShojikuLogger logger = message -> {};
    Map<String, SigningProvider> providers = Map.of("p", signer());

    Config merged =
        Config.empty()
            .merge(
                Config.empty()
                    .withTemplates("/t")
                    .withFontDirs(List.of("/f"))
                    .withLocaleDirs(List.of("/l"))
                    .withLang("ja-JP")
                    .withLibrary("/lib.so")
                    .withLogger(logger)
                    .withStrict(true)
                    .withProviders(providers)
                    .withEnv(false));

    assertEquals("/t", merged.templates());
    assertEquals(List.of("/f"), merged.fontDirs());
    assertEquals(List.of("/l"), merged.localeDirs());
    assertEquals("ja-JP", merged.lang());
    assertEquals("/lib.so", merged.library());
    assertSame(logger, merged.logger());
    assertTrue(merged.isStrict());
    assertSame(providers, merged.providers());
    assertFalse(merged.envEnabled());
  }

  @Test
  void anAbsentOverrideLeavesTheDefaultAlone() {
    Config configured =
        Config.empty()
            .withTemplates("/t")
            .withFontDirs(List.of("/f"))
            .withLocaleDirs(List.of("/l"))
            .withLang("ja-JP")
            .withLibrary("/lib.so")
            .withEnv(false);

    Config merged = configured.merge(Config.empty());

    assertEquals("/t", merged.templates());
    assertEquals(List.of("/f"), merged.fontDirs());
    assertEquals(List.of("/l"), merged.localeDirs());
    assertEquals("ja-JP", merged.lang());
    assertEquals("/lib.so", merged.library());
    assertFalse(merged.envEnabled());
  }

  @Test
  void mergingProducesACopyRatherThanMutatingTheDefaults() {
    Config configured = Config.empty().withTemplates("/t");

    configured.merge(Config.empty().withTemplates("/other"));

    assertEquals("/t", configured.templates());
  }

  @Test
  void aPerClientLangIsWhatARenderDefaultsTo() {
    assertTrue(client().lang("ja-JP").build().generate("receipt", null).success());
  }

  @Test
  void environmentLookupsAreOnByDefault() {
    assertTrue(Config.empty().envEnabled());
  }
}
